### PR TITLE
feat: add ec2 region provider

### LINKED ISSUE #320: detect aws region from ec2 metadata
Detect aws region from EC2 metadata service (when running on EC2 only of course) and integrate into: regionDetection

### PR DESCRIPTION
**Pull request by kneekey23**
## Issue \#

This closes #320 

## Conventional Commits

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AWSClientRuntime/Sources/Regions/DefaultRegionResolver.swift [hunk 1]
```diff
@@ -4,19 +4,24 @@
...
     }
     
     public func resolveRegion() -> String? {
         for provider in providers {
-            if let region = provider.resolveRegion() {
-                return region
+            do {
+                if let region = try provider.resolveRegion().get() {
+                    return region
+                }
```
Comment (wooj2) [inline AWSClientRuntime/Sources/Regions/DefaultRegionResolver.swift:22]: Add comment:
// TODO: Verify that this does not block calling thread.  If it does, we need to solve this similar to the way we solve this for the credentials provider

File: AWSClientRuntime/Sources/Regions/IMDSRegionProvider.swift [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        
+public struct IMDSRegionProvider: RegionProvider {
+    private let REGION_PATH = "/latest/meta-data/placement/region"
+    let imdsClient: IMDSClient?
+    
+    public init() {
+        do {
+            self.imdsClient = try IMDSClient()
+        } catch {
+            self.imdsClient = nil
```
Comment (wooj2) [inline AWSClientRuntime/Sources/Regions/IMDSRegionProvider.swift:17]: Add `log.err()` if we fail to instantiate IMDSClient()

File: AWSClientRuntime/Sources/Regions/IMDSRegionProvider.swift [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        }
+    }
+    
+    public func resolveRegion() -> Future<String?> {
+        let future = Future<String?>()
+        guard let imdsClient = imdsClient else {
+            future.fulfill(nil)
```
Comment (wooj2) [inline AWSClientRuntime/Sources/Regions/IMDSRegionProvider.swift:24]: Add `log.err()` if we fail to instantiate IMDSClient()

File: AWSClientRuntime/Sources/Regions/ProfileRegionResolver.swift [hunk 4]
```diff
@@ -22,11 +22,14 @@ public struct ProfileRegionProvider: RegionProvider {
...
-    public func resolveRegion() -> String? {
+    public func resolveRegion() -> Future<String?> {
+        let future = Future<String?>()
         guard let profile = profileCollection.profile(for: profileName) else {
-            return nil
+            future.fulfill(nil)
+            return future
         }
-        return profile.getProperty(name: "region")
+        future.fulfill(profile.getProperty(name: "region"))
```
Comment (wooj2) [inline AWSClientRuntime/Sources/Regions/ProfileRegionResolver.swift:31]:
```
guard let region = profile.getProperty(name: "region") else {
    log.info("Attempting to use \(profileName), but no region was found")
    future.fulfill(nil)
}
future.fulfill(region)
```

**Review by wooj2 — APPROVED** [review]
LGTM

File: AWSClientRuntime/Sources/Regions/DefaultRegionResolver.swift (hunk 1 above)
Comment (kneekey23) [reply AWSClientRuntime/Sources/Regions/DefaultRegionResolver.swift:22]: added

File: AWSClientRuntime/Sources/Regions/IMDSRegionProvider.swift (hunk 2 above)
Comment (kneekey23) [reply AWSClientRuntime/Sources/Regions/IMDSRegionProvider.swift:17]: not needed anymore as we no longer try to instantiate the client

File: AWSClientRuntime/Sources/Regions/IMDSRegionProvider.swift (hunk 3 above)
Comment (kneekey23) [reply AWSClientRuntime/Sources/Regions/IMDSRegionProvider.swift:24]: removed as we no longer have an optional client

File: AWSClientRuntime/Sources/Regions/ProfileRegionResolver.swift (hunk 4 above)
Comment (kneekey23) [reply AWSClientRuntime/Sources/Regions/ProfileRegionResolver.swift:31]: updated

File: AWSClientRuntime/Sources/Regions/RegionResolver.swift [hunk 5]
```diff
@@ -4,6 +4,7 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 //
+import AwsCommonRuntimeKit
```
[the remark this answers is not in the document]
Comment (kneekey23) [reply AWSClientRuntime/Sources/Regions/RegionResolver.swift:7]: removed

**Pull request merged**